lib/gc/Transforms/GPU/GpuUtils.h

```diff
@@ -0,0 +1,241 @@
+  }
+
+  int64_t getWorkGroupSize(Builder &builder) {
+    return getGpuPropertyAsInt(builder, "max_work_group_size",
+                               static_cast<DerivedT *>(this)->workGroupSize);
+  }
+};
+
+struct OpRewriter final : IRRewriter {
+  Location loc;
+
+  explicit OpRewriter(func::FuncOp &func)
+      : IRRewriter(func.getContext()), loc(func.getLoc()), func(func) {}
+
+  template <typename OpTy, typename... Args> OpTy create(Args &&...args) {
+    return RewriterBase::create<OpTy>(loc, std::forward<Args>(args)...);
+  }
+
+  arith::ConstantIndexOp getConstant(int64_t v) {
+    if (func.empty()) {
+      func.addEntryBlock();
+    } else {
+      // Try to find ConstantIndexOp with the same value in the function.
+      for (auto &op : func.getBody().getOps()) {
+        if (auto constOp = dyn_cast<arith::ConstantIndexOp>(op)) {
+          if (constOp.value() == v) {
+            return constOp;
+          }
+        }
+      }
+    }
+
+    auto ip = saveInsertionPoint();
+    setInsertionPointToStart(&func.getBody().front());
+    auto op = create<arith::ConstantIndexOp>(v);
+    restoreInsertionPoint(ip);
+    return op;
+  }
+
+private:
+  func::FuncOp &func;
+};
+
+// Round to the largest power of 2 that is <= value.
+template <typename T> static T floorPow2(T value) {
+  auto v = static_cast<std::make_unsigned_t<T>>(value);
+  return T(1) << llvm::bit_width(v) - 1;
+}
+
+// Round to the smallest power of 2 that is >= value.
+template <typename T> static T ceilPow2(T value) {
+  auto v = static_cast<std::make_unsigned_t<T>>(value);
+  return llvm::bit_ceil(v);
+}
+
+// Adjust tile sizes that meet the following conditions:
+// 1. The product of all tiles is as close to totalSize as possible.
+// 2. The new sizes are proportional to the initial sizes.
+// 3. The new sizes are powers of 2.
+template <typename T>
+static void adjustTiles(T totalSize, T *begin, T *end,
+                        T minSize = static_cast<T>(1)) {
+  assert((minSize & (minSize - 1)) == 0 && "minSize must be a power of 2");
+  auto count = end - begin;
+  if (count == 0) {
+    return;
+  }
+
+  T total = ceilPow2(totalSize);
+
+  if (count == 1) {
+    *begin = std::min(std::max(ceilPow2(*begin), minSize), total);
+    return;
+  }
+
+  if (count > 2) {
+    // Split the array in two. The first one consists of the 2 elements - the
+    // first one and the product of the rest. The second one is the rest.
+    T first[] = {*begin, std::accumulate(begin + 2, end, *(begin + 1),
+                                         std::multiplies<>())};
+    adjustTiles(total, first, first + 2, minSize);
+    adjustTiles(total / *first, begin + 1, end, minSize);
+    *begin = *first;
+    return;
+  }
+
+  --end;
+  T a = *begin;
+  T b = *end;
+  bool swap = a < b;
+  if (swap) {
+    std::swap(a, b);
+  }
+
+  a = ceilPow2(a);
+  b = floorPow2(b);
+
+  if (a * b <= total) {
+    *begin = swap ? b : a;
+    *end = swap ? a : b;
```

Comment:
There is `--end` above and we write the results inplace, into the input array. Do you mean it would be better to create new variables for better readability?